pyccl/halos/concentration.py

```diff
@@ -0,0 +1,308 @@
+        # Using central finite differences
+        lk_hi = lk_R + 0.005
+        lk_lo = lk_R - 0.005
+        dlpk = np.log(linear_matter_power(cosmo, np.exp(lk_hi), a) /
+                      linear_matter_power(cosmo, np.exp(lk_lo), a))
+        dlk = lk_hi - lk_lo
+        n = dlpk / dlk
+
+        sig = sigmaM(cosmo, M_use, a)
+        delta_c = 1.68647
+        nu = delta_c / sig
+
+        floor = self.phi_0 + n * self.phi_1
+        nu0 = self.eta_0 + n * self.eta_1
+        c = 0.5 * floor * ((nu0 / nu)**self.alpha +
+                           (nu / nu0)**self.beta)
+        if np.ndim(M) == 0:
+            c = c[0]
+
+        return c
+
+
+class ConcentrationBhattacharya13(Concentration):
+    name = 'Bhattacharya13'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationBhattacharya13, self).__init__(mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _check_mdef(self, mdef):
+        if mdef.Delta != 'vir':
+            if np.fabs(mdef.Delta - 200.) > 1E-4:
+                return True
+        return False
+
+    def _setup(self):
+        if self.mdef.Delta == 'vir':
+            self.A = 7.7
+            self.B = 0.9
+            self.C = -0.29
+        else:  # Now Delta has to be 200
+            if self.mdef.rho_type == 'matter':
+                self.A = 9.0
+                self.B = 1.15
+                self.C = -0.29
+            else:  # Now rho_type has to be critical
+                self.A = 5.9
+                self.B = 0.54
+                self.C = -0.35
+
+    def concentration(self, cosmo, M, a):
+        gz = growth_factor(cosmo, a)
+        status = 0
+        delta_c, status = lib.dc_NakamuraSuto(cosmo.cosmo, a, status)
+        sig = sigmaM(cosmo, M, a)
+        nu = delta_c / sig
+        return self.A * gz**self.B * nu**self.C
+
+
+class ConcentrationPrada12(Concentration):
+    name = 'Prada12'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationPrada12, self).__init__(mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _check_mdef(self, mdef):
+        if (np.fabs(mdef.Delta - 200.) > 1E-4) and \
+           (mdef.rho_type != 'critical'):
+            return True
+        return False
+
+    def _setup(self):
+        self.c0 = 3.681
+        self.c1 = 5.033
+        self.al = 6.948
+        self.x0 = 0.424
+        self.i0 = 1.047
+        self.i1 = 1.646
+        self.be = 7.386
+        self.x1 = 0.526
+        self.cnorm = 1. / self.cmin(1.393)
+        self.inorm = 1. / self.imin(1.393)
+
+    def cmin(self, x):
+        return self.c0 + (self.c1 - self.c0) * \
+            (np.arctan(self.al * (x - self.x0)) / np.pi + 0.5)
+
+    def imin(self, x):
+        return self.i0 + (self.i1 - self.i0) * \
+            (np.arctan(self.be * (x - self.x1)) / np.pi + 0.5)
+
+    def concentration(self, cosmo, M, a):
+        sig = sigmaM(cosmo, M, a)
+        om = cosmo.cosmo.params.Omega_m
```

Comment:
This will be OmegaM with massive neutrinos if they are non-zero. IDK if we care or not. Simply pointing it out.

Reply:
I think none of these have been calibrated with neutrinos, but that doesn't mean that we should exclude neutrinos here. @tmcclintock may know!

Reply:
I agree with both of your statements.